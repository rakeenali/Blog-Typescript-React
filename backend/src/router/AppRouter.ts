import express from "express";

export default class AppRouter {
  static instance: express.Router;

  static getRouter(): express.Router {
    if (!AppRouter.instance) {
      AppRouter.instance = express.Router();
    }

    return AppRouter.instance;
  }
}
