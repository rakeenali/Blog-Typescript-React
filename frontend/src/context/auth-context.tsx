import * as React from "react";
import { navigate } from "@reach/router";

import { Dispatch, ContextProps } from "../types/types";

export interface IAuthState {
  isAuthenticated: boolean;
  token: string;
}

type Action = { type: "LOGOUT" } | { type: "LOGIN"; payload: string };

export const AuthStateContext = React.createContext<IAuthState | null>(null);
export const AuthDispatchContext = React.createContext<Dispatch<Action> | null>(
  null
);

const authReducer = (state: IAuthState, action: Action) => {
  switch (action.type) {
    case "LOGIN": {
      return { ...state, isAuthenticated: true, token: action.payload };
    }
    case "LOGOUT": {
      return { ...INITIALSTATE };
    }
    default: {
      throw new Error(`Action type is not found in user context`);
    }
  }
};

const INITIALSTATE: IAuthState = {
  isAuthenticated: false,
  token: ""
};

export const AuthProvider = ({ children }: ContextProps): JSX.Element => {
  const [state, dispatch] = React.useReducer(authReducer, INITIALSTATE);

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  );
};

export function useAuthState(): IAuthState {
  const Context = React.useContext(AuthStateContext);

  if (!Context) {
    throw new Error(`useAuthState must be used iniside auth provider`);
  }

  return Context;
}

export function useAuthDispatch(): Dispatch<Action> {
  const Context = React.useContext(AuthDispatchContext);

  if (!Context) {
    throw new Error(`useAuthDispatch must be used iniside auth provider`);
  }

  return Context;
}

export const setUserAction = (
  token: string,
  dispatch: Dispatch<Action>
): void => {
  navigate("/profile");
  localStorage.setItem("blog:key", JSON.stringify(token));
  dispatch({ type: "LOGIN", payload: token });
};

export const getUserAction = (dispatch: Dispatch<Action>): void => {
  const rawToken = localStorage.getItem("blog:key");
  if (rawToken) {
    const token: string = JSON.parse(rawToken);
    dispatch({ type: "LOGIN", payload: token });
  }
};

export const logout = (dispatch: Dispatch<Action>): void => {
  localStorage.removeItem("blog:key");
  navigate("/login");
  dispatch({ type: "LOGOUT" });
};
