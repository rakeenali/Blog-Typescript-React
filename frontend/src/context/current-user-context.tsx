import * as React from "react";
import { CurrentUserResponse as CurrentUserState } from "../types/inteface";
import { useAuthState } from "./auth-context";
import axios, { AxiosResponse } from "axios";
import { Dispatch, ContextProps } from "../types/types";

type Action = { type: "SET_USER"; payload: CurrentUserState };

export const CurrentUserStateContext = React.createContext<CurrentUserState | null>(
  null
);
export const CurrentUserDispatchContext = React.createContext<Dispatch<
  Action
> | null>(null);

const currentUserReducer = (state: CurrentUserState, action: Action) => {
  switch (action.type) {
    case "SET_USER": {
      return {
        ...state,
        ...action.payload
      };
    }
    default: {
      throw new Error(`Action type is not found in user context`);
    }
  }
};

const INITIALSTATE: CurrentUserState = {
  userId: null,
  name: null,
  username: null,
  usersProfile: null,
  profile: null,
  blogs: []
};

export const CurrentUserProvider = ({
  children
}: ContextProps): JSX.Element => {
  const [state, dispatch] = React.useReducer(currentUserReducer, INITIALSTATE);
  const authState = useAuthState();

  React.useLayoutEffect(() => {
    if (authState.isAuthenticated) {
      axios
        .get("http://localhost:5000/profile/current", {
          headers: { authorization: authState.token }
        })
        .then((res: AxiosResponse<CurrentUserState>) => {
          dispatch({ type: "SET_USER", payload: res.data });
        });
    }
  }, [authState]);

  return (
    <CurrentUserStateContext.Provider value={state}>
      <CurrentUserDispatchContext.Provider value={dispatch}>
        {children}
      </CurrentUserDispatchContext.Provider>
    </CurrentUserStateContext.Provider>
  );
};

export function useCurrentUserState(): CurrentUserState {
  const Context = React.useContext(CurrentUserStateContext);

  if (!Context) {
    throw new Error(
      `useCurrentUserState must be used iniside CurreuntUser provider`
    );
  }

  return Context;
}

export function useCurrentUserDispatch(): Dispatch<Action> {
  const Context = React.useContext(CurrentUserDispatchContext);

  if (!Context) {
    throw new Error(
      `useCurrentUserDispatch must be used iniside CurreuntUser provider`
    );
  }

  return Context;
}

export const getProfile = (token: string, dispatch: Dispatch<Action>): void => {
  axios
    .get("http://localhost:5000/profile/current", {
      headers: { authorization: token }
    })
    .then((res: AxiosResponse<CurrentUserState>) => {
      dispatch({ type: "SET_USER", payload: res.data });
    });
};
