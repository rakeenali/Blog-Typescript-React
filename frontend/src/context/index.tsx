import React from "react";

import { AuthProvider } from "./auth-context";

type AppProps = { children: React.ReactNode };

export default function AppProvider({ children }: AppProps): JSX.Element {
  return <AuthProvider>{children}</AuthProvider>;
}
