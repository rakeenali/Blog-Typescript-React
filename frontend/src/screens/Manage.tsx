import React from "react";
import styled from "styled-components";
import { Card, FormButton } from "../theme";
import axios from "axios";
import Error from "../components/Error";
import { SubmitProfile as IFormInput } from "../types/inteface";
import { ErrorResponse, OKResponse } from "../types/types";
import { navigate } from "@reach/router";

import {
  useCurrentUserState,
  useCurrentUserDispatch,
  getProfile
} from "../context/current-user-context";
import { useAuthState } from "../context/auth-context";

const Manage = (): JSX.Element => {
  const userState = useCurrentUserState();
  const userDispatch = useCurrentUserDispatch();
  const authState = useAuthState();

  const [showError, setShowError] = React.useState<boolean>(false);
  const [errorMessage, setErrorMessage] = React.useState<string>("");
  const [text, setText] = React.useState<IFormInput>({
    firstName: "",
    lastName: "",
    avatar: ""
  });

  React.useMemo(() => {
    if (userState.profile) {
      setText({
        firstName: userState.profile.firstName,
        lastName: userState.profile.lastName,
        avatar: userState.profile.avatar
      });
    }
  }, [userState]);

  const closeError = () => {
    setShowError(false);
    setErrorMessage("");
  };

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setText({ ...text, [e.target.name]: e.target.value });
  };

  const manageProfile = (e: React.FormEvent) => {
    e.preventDefault();

    axios
      .post(
        "http://localhost:5000/profile/manage",
        {
          firstName: text.firstName,
          lastName: text.lastName,
          avatar: text.avatar
        },
        { headers: { authorization: authState.token } }
      )
      .then((res: OKResponse) => {
        if (res && res.data.message) {
          navigate("/profile");
          getProfile(authState.token, userDispatch);
        }
      })
      .catch((err: ErrorResponse) => {
        if (err && err.response) {
          setShowError(true);
          setErrorMessage(err.response.data.error);
        }
      });
  };

  return (
    <Center>
      {showError && <Error message={errorMessage} closeError={closeError} />}
      <Card height="80%" width="60%">
        <Inner>
          <Header>Manage Profile</Header>
          <div>
            <Form onSubmit={manageProfile}>
              <FormControl>
                <label htmlFor="firstName">First Name</label>
                <input
                  autoFocus
                  autoComplete="off"
                  type="text"
                  id="firstName"
                  placeholder="First Name"
                  onChange={onChange}
                  value={text.firstName}
                  required
                  name="firstName"
                />
              </FormControl>
              <FormControl>
                <label htmlFor="lastName">Last Name</label>
                <input
                  autoComplete="off"
                  type="text"
                  id="lastName"
                  placeholder="Last Name"
                  onChange={onChange}
                  value={text.lastName}
                  required
                  name="lastName"
                />
              </FormControl>
              <FormControl>
                <label htmlFor="avatar">Avatar's Link</label>
                <input
                  autoComplete="off"
                  type="text"
                  id="avatar"
                  placeholder="Avatar's Link"
                  onChange={onChange}
                  value={text.avatar}
                  required
                  name="avatar"
                />
              </FormControl>
              <FormControl>
                <FormButton>Proceed</FormButton>
              </FormControl>
            </Form>
          </div>
        </Inner>
      </Card>
    </Center>
  );
};

const Center = styled.div`
  width: 100%;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Inner = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px;
`;

const Header = styled.h1`
  flex-basis: 20%;
  font-size: 50px;
`;
const Form = styled.form`
  flex-basis: 80%;
  height: 100%;
  width: 100%;
`;

const FormControl = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  margin-bottom: 20px;

  label {
    font-size: 20px;
    margin-bottom: 10px;
    text-align: center;
  }

  input {
    outline-color: red;
    padding: 10px;
    width: 35vw;
    background-color: #ccc;
    border-radius: 8px;
    border: none;
    font-size: 17px;
  }
`;

export default Manage;

// https://images.pexels.com/photos/67636/rose-blue-flower-rose-blooms-67636.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500
