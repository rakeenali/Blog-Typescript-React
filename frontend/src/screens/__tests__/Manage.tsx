import "@testing-library/jest-dom/extend-expect";
import * as React from "react";
import {
  render,
  fireEvent,
  wait,
  waitForDomChange
} from "@testing-library/react";
import axios from "axios";
import { navigate } from "@reach/router";

import { getUser, getProfile } from "../../test/buildUser";
import * as AuthContext from "../../context/auth-context";
import * as CurrentUserContext from "../../context/current-user-context";
import Manage from "../Manage";

jest.mock("@reach/router", () => {
  const navigate = jest.fn();
  return { navigate };
});

beforeEach(() => {
  jest.spyOn(axios, "post").mockImplementation((...args) => {
    console.warn("axios.post is not mocked for this call", ...args);
    return Promise.reject(new Error("This must be mocked!"));
  });
});

afterEach(() => {
  const postAxios = axios.post as jest.Mock;
  const navigateMock = navigate as jest.Mock;
  postAxios.mockClear();
  navigateMock.mockClear();
});

describe("<Manage/>", () => {
  test("renders the form correctly", async () => {
    const { container } = render(
      <AuthContext.AuthProvider>
        <CurrentUserContext.CurrentUserProvider>
          <Manage />
        </CurrentUserContext.CurrentUserProvider>
      </AuthContext.AuthProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("handle form changes correctly", () => {
    const firstNameText = "Rakeen";
    const lastNameText = "Ali";
    const avatarUrl =
      "https://images.pexels.com/photos/67636/rose-blue-flower-rose-blooms-67636.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500";

    const { getByPlaceholderText } = render(
      <AuthContext.AuthProvider>
        <CurrentUserContext.CurrentUserProvider>
          <Manage />
        </CurrentUserContext.CurrentUserProvider>
      </AuthContext.AuthProvider>
    );

    const firstName = getByPlaceholderText("First Name") as HTMLInputElement;
    const lastName = getByPlaceholderText("Last Name") as HTMLInputElement;
    const avatarLink = getByPlaceholderText(
      "Avatar's Link"
    ) as HTMLInputElement;

    fireEvent.change(firstName, { target: { value: firstNameText } });
    fireEvent.change(lastName, { target: { value: lastNameText } });
    fireEvent.change(avatarLink, { target: { value: avatarUrl } });

    expect(firstName.value).toEqual(firstNameText);
    expect(lastName.value).toEqual(lastNameText);
    expect(avatarLink.value).toEqual(avatarUrl);
  });

  test("can add profile", async () => {
    CurrentUserContext.getProfile = jest.fn();
    const axiosPost = axios.post as jest.Mock;

    const firstNameText = "Rakeen";
    const lastNameText = "Ali";
    const avatarUrl =
      "https://images.pexels.com/photos/67636/rose-blue-flower-rose-blooms-67636.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500";

    const { getByPlaceholderText, queryByText } = render(
      <AuthContext.AuthProvider>
        <CurrentUserContext.CurrentUserProvider>
          <Manage />
        </CurrentUserContext.CurrentUserProvider>
      </AuthContext.AuthProvider>
    );

    const firstName = getByPlaceholderText("First Name") as HTMLInputElement;
    const lastName = getByPlaceholderText("Last Name") as HTMLInputElement;
    const avatarLink = getByPlaceholderText(
      "Avatar's Link"
    ) as HTMLInputElement;

    fireEvent.change(firstName, { target: { value: firstNameText } });
    fireEvent.change(lastName, { target: { value: lastNameText } });
    fireEvent.change(avatarLink, { target: { value: avatarUrl } });

    expect(firstName.value).toEqual(firstNameText);
    expect(lastName.value).toEqual(lastNameText);
    expect(avatarLink.value).toEqual(avatarUrl);

    axiosPost.mockResolvedValueOnce({
      data: {
        message: "Profile updated"
      }
    });

    const proceedButton = queryByText("Proceed");
    proceedButton && fireEvent.click(proceedButton);

    await wait(
      () => {
        expect(navigate).toHaveBeenCalledTimes(1);
        expect(CurrentUserContext.getProfile).toHaveBeenCalledTimes(1);
      },
      { timeout: 100 }
    );
  });

  test("can update profile", async () => {
    CurrentUserContext.getProfile = jest.fn();
    const axiosPost = axios.post as jest.Mock;

    const profile = getProfile();
    const user = getUser({ profile });

    const { getByPlaceholderText, queryByText } = render(
      <AuthContext.AuthProvider>
        <CurrentUserContext.CurrentUserStateContext.Provider value={user}>
          <CurrentUserContext.CurrentUserDispatchContext.Provider
            value={() => null}
          >
            <Manage />
          </CurrentUserContext.CurrentUserDispatchContext.Provider>
        </CurrentUserContext.CurrentUserStateContext.Provider>
      </AuthContext.AuthProvider>
    );

    const firstName = getByPlaceholderText("First Name") as HTMLInputElement;
    const lastName = getByPlaceholderText("Last Name") as HTMLInputElement;
    const avatarLink = getByPlaceholderText(
      "Avatar's Link"
    ) as HTMLInputElement;

    expect(firstName.value).toEqual(profile.firstName);
    expect(lastName.value).toEqual(profile.lastName);
    expect(avatarLink.value).toEqual(profile.avatar);

    fireEvent.change(firstName, { target: { value: "Rakeen" } });
    expect(firstName.value).toEqual("Rakeen");

    axiosPost.mockResolvedValueOnce({
      data: {
        message: "Profile updated"
      }
    });

    const proceedButton = queryByText("Proceed");

    proceedButton && fireEvent.click(proceedButton);

    await wait(
      () => {
        expect(navigate).toHaveBeenCalledTimes(1);
        expect(CurrentUserContext.getProfile).toHaveBeenCalledTimes(1);
      },
      { timeout: 100 }
    );
  });

  test("can render error correctly", async () => {
    const axiosPost = axios.post as jest.Mock;

    const firstNameText = "Rakeen";
    const avatarUrl =
      "https://images.pexels.com/photos/67636/rose-blue-flower-rose-blooms-67636.jpeg?auto=compress&cs=tinysrgb&dpr=1&w=500";

    const { getByPlaceholderText, debug, queryByText } = render(
      <AuthContext.AuthProvider>
        <CurrentUserContext.CurrentUserProvider>
          <Manage />
        </CurrentUserContext.CurrentUserProvider>
      </AuthContext.AuthProvider>
    );

    const firstName = getByPlaceholderText("First Name") as HTMLInputElement;
    const avatarLink = getByPlaceholderText(
      "Avatar's Link"
    ) as HTMLInputElement;

    fireEvent.change(firstName, { target: { value: firstNameText } });
    fireEvent.change(avatarLink, { target: { value: avatarUrl } });

    const proceedButton = queryByText("Proceed");
    const errorMessage = "Last Name field is required";

    axiosPost.mockRejectedValueOnce({
      response: {
        data: {
          error: errorMessage
        }
      }
    });

    proceedButton && fireEvent.click(proceedButton);

    await waitForDomChange();

    expect(queryByText(errorMessage)).toBeInTheDocument();
  });
});
