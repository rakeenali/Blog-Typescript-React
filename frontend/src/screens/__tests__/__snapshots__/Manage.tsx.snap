// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Manage/> renders the form correctly 1`] = `
<div
  class="sc-ifAKCX kMXFrU"
>
  <div
    class="sc-bdVaJa ilTCYI"
    height="80%"
    width="60%"
  >
    <div
      class="sc-EHOje ePbsYW"
    >
      <h1
        class="sc-bZQynM hQqxvj"
      >
        Manage Profile
      </h1>
      <div>
        <form
          class="sc-gzVnrw ieXoIT"
        >
          <div
            class="sc-htoDjs khaQSc"
          >
            <label
              for="firstName"
            >
              First Name
            </label>
            <input
              autocomplete="off"
              id="firstName"
              name="firstName"
              placeholder="First Name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs khaQSc"
          >
            <label
              for="lastName"
            >
              Last Name
            </label>
            <input
              autocomplete="off"
              id="lastName"
              name="lastName"
              placeholder="Last Name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs khaQSc"
          >
            <label
              for="avatar"
            >
              Avatar's Link
            </label>
            <input
              autocomplete="off"
              id="avatar"
              name="avatar"
              placeholder="Avatar's Link"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs khaQSc"
          >
            <button
              class="sc-bwzfXH ktHgFu"
            >
              Proceed
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
